granulate_utils/linux/ns.py

```diff
@@ -282,3 +282,18 @@ def resolve_host_path(process: Process, ns_path: str) -> str:
     Get a path in the host mount namespace pointing to path in process mount namespace.
     """
     return resolve_proc_root_links(get_proc_root_path(process), ns_path)
+
+
+def get_host_pid(nspid: int, container_id: str) -> Optional[int]:
+    assert len(container_id) == 64, f"Invalid container id {container_id!r}"
+    pid_namespace = ""
+    for process in process_iter():
+        try:
+            if not pid_namespace and container_id in Path(f"/proc/{process.pid}/cgroup").read_text():
+                pid_namespace = os.readlink(f"/proc/{process.pid}/ns/pid")
```

Comment:
Technically, after starting, you need to re-start the loop: we want to catch processes that run in the PID NS but don't share the cgroup, each processes started by `nsenter` or `run_in_ns`.